bridges/snowbridge/pallets/outbound-queue-v2/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+//! Pallet for committing outbound messages for delivery to Ethereum
+//!
+//! # Overview
+//!
+//! Messages come either from sibling parachains via XCM, or BridgeHub itself
+//! via the `snowbridge-pallet-system-v2`:
+//!
+//! 1. `snowbridge_outbound_queue_primitives::v2::EthereumBlobExporter::deliver`
+//! 2. `snowbridge_pallet_system_v2::Pallet::send`
+//!
+//! The message submission pipeline works like this:
+//! 1. The message is first validated via the implementation for
+//!    [`snowbridge_outbound_queue_primitives::v2::SendMessage::validate`]
+//! 2. The message is then enqueued for later processing via the implementation for
+//!    [`snowbridge_outbound_queue_primitives::v2::SendMessage::deliver`]
+//! 3. The underlying message queue is implemented by [`Config::MessageQueue`]
+//! 4. The message queue delivers messages to this pallet via the implementation for
+//!    [`frame_support::traits::ProcessMessage::process_message`]
+//! 5. The message is processed in `Pallet::do_process_message`:
+//! 	a. Convert to `OutboundMessage`, and stored into the `Messages` vector storage
+//! 	b. ABI-encode the `OutboundMessage` and store the committed hash in `MessageLeaves`
+//! 	c. Generate `PendingOrder` with assigned nonce and fee attached, stored into the
+//! `PendingOrders` 	   map storage, with nonce as the key
+//! 	d. Increment nonce and update the `Nonce` storage
+//! 6. At the end of the block, a merkle root is constructed from all the leaves in `MessageLeaves`,
+//!    then `MessageLeaves` is dropped so that it is never committed to storage or included in PoV.
```

Comment:
The code does not drop `MessageLeaves` at the end of the block, so it does in fact get included in storage and PoV.

```suggestion
//! 6. At the end of the block, a merkle root is constructed from all the leaves in `MessageLeaves`.
//!    then `MessageLeaves` is dropped so that it is never committed to storage or included in PoV.
//!    At the beginning of the next block, both `Messages` and `MessageLeaves` are dropped so that
//!    state at each block only holds the messages processed in that block.
```

Why this design though? Why not keep messages in storage and their hashes in MerkleTree up until the message delivery proof comes back and then prune?

What is more expensive to process on-chain on Ethereum?:
- walking back the BH blocks from closest BEEFY anchor/proof, back to the block that includes the message,
  or,
- use the exact BH block from the closest BEEFY anchor/proof, but then verify a larger merkle proof of `MessageLeaves`?

I expect the latter, so the current design looks like it's not optimizing for cost on eth (largest cost afaik), so what's the reasoning?

Reply:
Opened https://github.com/Snowfork/snowbridge/issues/1408 for follow-up